hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -2203,6 +2204,21 @@ public Optional<Boolean> shouldCacheFile(HFileInfo hFileInfo, Configuration conf
     return Optional.of(!fullyCachedFiles.containsKey(fileName));
   }
 
+  @Override
+  public Optional<Boolean> shouldCacheBlock(BlockCacheKey key, Configuration conf) {
+    try {
+      DataTieringManager dataTieringManager = DataTieringManager.getInstance();
+      if (dataTieringManager != null && !dataTieringManager.isHotData(key, conf)) {
```

Comment:
@vinayakphegde, @wchevreuil,
We are only transiently using the timestamp within BlockCacheKey in this API. But, we are keeping this value permanently in the BlockCacheKey increasing the size of the BlockCacheKey and keeping the timestamp redundantly, which will not be used later. This overhead can be avoided by passing the required data as arguments to this function shouldCacheBlock.
All we need is this type of a function:
`shouldCacheBlock(long timestamp, long age or conf)`
wouldn't it make sense to not increase the size of BlockCacheKey permanently?

Reply:
All we need is a call to
`hotDataValidator(maxTimestamp, getDataTieringHotDataAge(configuration));` to decide whether to cache the block or not. We already have the maxTimestamp and age handy at the caller of shouldCacheBlock.

Reply:
I agree, it will increase the size of each BlockCacheKey by 8 bytes. Assuming we have 10 million BlockCacheKeys on average on each region server (I think, this is what @wchevreuil said the other day), this would increase the heap size by approximately 76 MB.

However, I believe the proposed solution is not correct. The `shouldCacheBlock` function should not process information related to data tiering; this responsibility should be delegated to the `DataTieringManager`. Additionally, we should not assume that the DataTieringType of the block is TIME_RANGE, as it could be different in the future.

Reply:
The general idea behind storing any data in the memory (as BlockCacheKey will reside in memory) is to use the data in future. Here, we are increasing the size of BlockCacheKey, transiently using the newly added member for making caching decisions, but later keeping the data in the BlockCacheKey without using it for later. Additionally this increase in size is effective even if the feature is not enabled, since we have changed the class layout.

Hence, I would propose to not store it in BlockCacheKey but store it on the stack (member variable) so that the variable will go away after getting used.

Reply:
Yeah, I understand the concern. The only thing I am worried about is how we can achieve it. Let me think about a way to achieve this without compromising on OO principles.

Reply:
On a file based bucket cache over an ephemeral storage of 1.6TB capacity and a typical block size of 64KB, we may get up to 26M blocks. With 4 extra bytes this will add per block, we are talking about 100MB extra heap usage, which accounts for 0.3% of a 32GB heap on typical RS deployment, so I guess we can relax on this extra variable.

WDYT, @jhungund ?

Reply:
Hi @wchevreuil, In my humble opinion, we usually code with the following order of priorities:

Functional correctness >  Performance/Resource utilisation > Coding guidelines/principles

We cannot let go of functional correctness for the sake of performance or for coding principles. 
Similarly, we may not want to relax the performance/resource utilisation for the sake of adhering to coding principles.

Here we are repurposing an object that stays in memory forever as an information carrier for our utility function (isHotData). Once the utility function finishes, we will never use that data again. 

Can we have a utility defined in DataTieringManager to take timestamps, configuration objects and make decisions for us?

Sorry, I do not want to hold on to this change by being inflexible, but trying to explain that we can achieve the same result by relaxing the OO coding principles instead of relaxing the memory utilisation.

@vinayakphegde and I will spend one more day and try to come up with a solution. If we can't come up with a solution, we can go ahead and submit this change.
Thanks,
Janardhan